File: week1/problemSet1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main (void) {

    // Prompt user for card number
    long cardNumber = get_long("Number: ");
    long lastDigit = cardNumber;

    int sum1 = 0;
    int sum2 = 0;

    int cardDigits;

    // Luhn's Algorithm
    for (cardDigits = 0; cardNumber > 0; cardDigits ++) {

        if (cardDigits % 2 != 0) {

            if ((cardNumber % 10) * 2 > 9) {
                sum1 = sum1 + (2 * (cardNumber % 10) - 9);
            } else
                sum1 = sum1 + ((cardNumber % 10) * 2);
        } else {
            sum2 = sum2 + cardNumber % 10;
        }

        cardNumber = cardNumber / 10;
    }

    // Validat if after Luhn's algorithm card number is correct
    if ((sum1 + sum2) % 10 == 0) {

        // Taking first 2 digits to check which bank card is from
        lastDigit = lastDigit / pow(10, (cardDigits - 2));


        // Asigning a bank depending on first 2 digits
        if (cardDigits == 13 && 39 < lastDigit && lastDigit < 50) {
            printf("VISA\n");
        } else if (cardDigits == 15 && (lastDigit == 34 || lastDigit == 37)) {
            printf("AMEX\n");
        } else if (cardDigits == 16 && 39 < lastDigit && lastDigit < 50) {
            printf("VISA\n");
        } else if (cardDigits == 16 && (lastDigit == 51 || lastDigit == 52 || lastDigit == 53 || lastDigit == 54 || lastDigit == 55)){
            printf("MASTERCARD\n");
        } else {
            printf("INVALID\n");
        }
    }
    else {
        printf("INVALID\n");
    }
}

File: week6/problemSet6/sentimental-credit/credit.py
# TODO
from cs50 import get_int

def main():

    cardNumber = get_int("Card Number: ")

    digits = digitsCounter(cardNumber)

    # Validation of AMEX card
    if digits == 15 and str(cardNumber)[:2] == "37":
        if luhns(cardNumber):
            print("AMEX")
        else:
            print("INVALID")

    # Validation of VISA card
    elif (digits == 13 or digits == 16) and str(cardNumber)[:1] == "4":
        if luhns(cardNumber):
             print("VISA")
        else:
            print("INVALID")

    # Validation of MASTERCARD card
    elif digits == 16 and (str(cardNumber)[:2] == "51" or str(cardNumber)[:2] == "52" or str(cardNumber)[:2] == "53" or str(cardNumber)[:2] == "54" or str(cardNumber)[:2] == "55"):
        if luhns(cardNumber):
             print("MASTERCARD")
        else:
            print("INVALID")
    else:
        print("INVALID")

def digitsCounter(numbers):
    digits = 1
    while numbers > 10:
        numbers = numbers / 10
        digits += 1
    return digits

def luhns(numbers):
    sum = 0
    for i, digit in enumerate(reversed(str(numbers))):
        if i % 2 == 0:
            sum += int(digit)
        else:
            double = int(digit) * 2
            sum += double // 10 + double % 10
    return sum % 10 == 0

main()